Remove stage_index_to_group_rank from schedule

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146217
* #146193

This PR allows schedules loaded via CSV to automatically set their `stage_index_to_group_rank ` and removes the `stage_index_to_group_rank ` argument from the `PipelineScheduleMulti` constructor

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
Remove stage_index_to_group_rank from schedule

[ghstack-poisoned]

H-Huang committed:
Update on "Remove stage_index_to_group_rank from schedule"




cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on torch/distributed/pipelining/schedules.py:
```diff
@@ -1124,12 +1132,9 @@ def __init__(
-            )
-        else:
-            self.stage_index_to_group_rank = stage_index_to_group_rank
+        self.stage_index_to_group_rank = generate_stage_to_rank_mapping(
```
i'm confused. are we first generating via generate_... and then overwriting inside _validate by computing the same thing a different way?

H-Huang replied:
Yeah, `_validate` is used when the schedule is loaded by CSV. There are two cases:

1. Using a default schedule (`Interleaved1F1B`, `BFS`, etc.)
- During initialize, `generate_stage_to_rank_mapping`
- pipeline actions are created based on the schedule implementation

2. Creating schedule to load with CSV
- During initialize, `generate_stage_to_rank_mapping` (this is not needed and eventually gets overridden)
- load csv to get the pipeline actions
- _validate which infers the `stage_to_rank_mapping` and sets it

I think to do this in a cleaner way I will need to probably need to make `PipelineScheduleMulti` an ABC and only allow the csv loading via the runtime class?

wconstab reviewed (approved):
I think this makes sense- it is logical to have a way to infer mapping from CSV and it is necessarily different code from the generate path. Just add a comment where you call generate that it's intentional to be overwritten in the specific CSV case.

H-Huang committed:
Update on "Remove stage_index_to_group_rank from schedule"


This PR allows schedules loaded via CSV to automatically set their `stage_index_to_group_rank ` and removes the `stage_index_to_group_rank ` argument from the `PipelineScheduleMulti` constructor

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorchbot merge -i
